### PR TITLE
bpo-39868: Add the documentation for Assignment Expressions (PEP 572).

### PR DESCRIPTION
**Pull request by shankarj67**
https://bugs.python.org/issue39868

https://bugs.python.org/issue39702

### CODE REVIEW

**Comment by brandtbucher** [comment]
Thanks for your time @shankarj67, and welcome to CPython! :sunglasses:

I'll review this in a bit. Adding "skip news" since the one for #18793 is fine here as well.

Also, can you change the PR title to reference "bpo-39868" instead of "bpo-39702"?

File: Doc/reference/expressions.rst [hunk 1]
```diff
@@ -1649,7 +1649,61 @@ Assignment expressions
...
+more readable and maintainable.
+
+The expression ``name := expr`` where **expr** is any valid Python expression
+other than an unparenthesize tuple and the **name** is an identifier.
+Here the variable **name** holds the assigned value of a given **expression**.
+
+**Examples**
+
+*The following examples demonstrate the use case of an assignment expressions.*
+
```
Comment (brandtbucher) [inline Doc/reference/expressions.rst]:
```suggestion
```

File: Doc/reference/expressions.rst [hunk 2]
```diff
@@ -1649,7 +1649,61 @@ Assignment expressions
...
+.. code-block:: python
+
+   matching = patern.search(data)
+   if matching:
+       do_something(matching)
+
+Notice that a matching variable is being called two times in the code
+that may be replaced using assignment expression in one line.
+
+**Using assignment expression**
```
Comment (brandtbucher) [inline Doc/reference/expressions.rst]:
```suggestion
```

File: Doc/reference/expressions.rst [hunk 3]
```diff
@@ -1649,7 +1649,61 @@ Assignment expressions
...
+**Current scenario**
+
+.. code-block:: python
+
+   chunk = file.read(9000)
+   while chunk:
+       process(chunk)
+       chunk = file.read(9000)
+
+**Using assignment expression**
```
Comment (brandtbucher) [inline Doc/reference/expressions.rst]:
```suggestion
```

**Review by brandtbucher — CHANGES_REQUESTED** [review]
This is good, thanks!

It's a bit longer than I think it needs to be, though (see similar sections, which are quite brief and have no subheadings). I also don't *think* we need examples of code *without* assignment expressions, since we're not trying to sell the syntax (like the PEP), just explaining it!

Before we get to grammar, etc., I've identified three chunks that can probably be removed, for the reasons above.

File: Doc/reference/expressions.rst (hunk 3 above)
Comment (shankarj67) [reply Doc/reference/expressions.rst]: I have added the commit.

File: Doc/reference/expressions.rst [hunk 4]
```diff
@@ -1649,7 +1649,32 @@ Assignment expressions
...
+Another popular use case is when processing the streams in a
+chunk from the file.
+
+.. code-block:: python
+
+   while chunk := file.read(9000):
+       process(chunk)
+
 
 See :pep:`572` for more details about assignment expressions.
```
Comment (brandtbucher) [inline Doc/reference/expressions.rst:1672]: We typically add `versionadded` directives for new features:

```suggestion

.. versionadded:: 3.8

```

File: Doc/reference/expressions.rst [hunk 5]
```diff
@@ -1649,7 +1649,32 @@ Assignment expressions
...
+
+One common use case is when handling the matched data in regular expression
+
+.. code-block:: python
+
+   if (matching := pattern.search(data)):
+       do_something(matching)
+
+Another popular use case is when processing the streams in a
+chunk from the file.
```
Comment (brandtbucher) [inline Doc/reference/expressions.rst]: I was a bit confused by the original wording... something like this is clearer to me:

```suggestion
Or, when processing a file stream in chunks:
```

File: Doc/reference/expressions.rst [hunk 6]
```diff
@@ -1649,7 +1649,32 @@ Assignment expressions
...
+The expression ``name := expr`` where **expr** is any valid Python expression
+other than an unparenthesize tuple and the **name** is an identifier.
+Here the variable **name** holds the assigned value of a given **expression**.
+
+
+One common use case is when handling the matched data in regular expression
+
+.. code-block:: python
+
+   if (matching := pattern.search(data)):
```
Comment (brandtbucher) [inline Doc/reference/expressions.rst]: No need for parentheses here:

```suggestion
   if matching := pattern.search(data):
```

File: Doc/reference/expressions.rst [hunk 7]
```diff
@@ -1649,7 +1649,32 @@ Assignment expressions
...
+which aims to reduce the number of lines of code by allowing the
+assignment and return of value in the same expression. This makes codebase
+more readable and maintainable.
+
+The expression ``name := expr`` where **expr** is any valid Python expression
+other than an unparenthesize tuple and the **name** is an identifier.
+Here the variable **name** holds the assigned value of a given **expression**.
+
+
+One common use case is when handling the matched data in regular expression
```
Comment (brandtbucher) [inline Doc/reference/expressions.rst]: Grammar:

```suggestion
One common use case is when handling matched regular expressions:
```

File: Doc/reference/expressions.rst [hunk 8]
```diff
@@ -1649,7 +1649,32 @@ Assignment expressions
...
+Assignment expression is the new expression introduced in the Python3.8
+or later (sometimes also called as "named expression" or "walrus operator"),
+which aims to reduce the number of lines of code by allowing the
+assignment and return of value in the same expression. This makes codebase
+more readable and maintainable.
+
+The expression ``name := expr`` where **expr** is any valid Python expression
+other than an unparenthesize tuple and the **name** is an identifier.
+Here the variable **name** holds the assigned value of a given **expression**.
+
```
Comment (brandtbucher) [inline Doc/reference/expressions.rst]: I personally don't think this needs to include the history or motivation... just the behavior. We can reuse the above grammar for the explanation, too!

```suggestion
An assignment expression (sometimes also called a "named expression" or
"walrus") assigns an :token:`expression` to an :token:`identifier`, while also
returning the value of the :token:`expression`. 
```

**Review by brandtbucher — CHANGES_REQUESTED** [review]
I've added some suggestions (mostly grammar and style):

**Comment by shankarj67** [comment]
Thank you for the suggestion. I have added the new commit.

File: Doc/reference/expressions.rst [hunk 9]
```diff
@@ -1649,9 +1649,26 @@ Assignment expressions
 .. productionlist::
    assignment_expression: [`identifier` ":="] `expression`
 
-.. TODO: BPO-39868
+An assignment expression (sometimes also called a "named expression" or
+"walrus") assigns an :token:`expression` to an :token:`identifier`, while also
+returning the value of the :token:`expression`.
 
-See :pep:`572` for more details about assignment expressions.
+One common use case is when handling matched regular expression:
```
Comment (brandtbucher) [inline Doc/reference/expressions.rst]: Thanks! One more tiny fix:

```suggestion
One common use case is when handling matched regular expressions:
```

Comment (shankarj67) [reply Doc/reference/expressions.rst]: done

**Review by brandtbucher — APPROVED** [review]

**Comment by brandtbucher** [comment]
Closing and reopening to trigger CI.

**Review by gvanrossum — APPROVED** [review]
This looks fine to me. Is it waiting for anything?

**Comment by brandtbucher** [comment]
Just a merge! ;)

**Pull request merged**